cam6_4_176: COSP-RTTOV integration with CAM (#1327)
ESCOMP/CAM: src/physics/cosp2/COSP-RTTOV_examples/README.txt

# Instructions on how to use COSP-RTTOV in CAM

> ### Jonah Shaw, 2026/02/18 (jonah.shaw@colorado.edu)

Reviewer: @jshaw35 given that you are leaving CU soon is this still an OK email for you?

Author: I should have a long-term CU affiliation that starts when I leave that allows me to use my email. Is that alright?

Reviewer: Yep, that's fine with us, I just wanted to double-check!  I'll go ahead and mark this comment as resolved.

Shaw, J. K., Swales, D. J., DeSouza-Machado, S., Turner, D. D., Kay, J. E., and Schneider, D. P.: COSP-RTTOV-1.0: flexible radiation diagnostics to enable new science applications in model evaluation, climate change detection, and satellite mission design, Geosci. Model Dev., 18, 4935–4950, https://doi.org/10.5194/gmd-18-4935-2025, 2025.

Swales, D. J., Pincus, R., and Bodas-Salcedo, A.: The Cloud Feedback Model Intercomparison Project Observational Simulator Package: Version 2, Geosci. Model Dev., 11, 77–81, https://doi.org/10.5194/gmd-11-77-2018, 2018.

Saunders, R., Hocking, J., Turner, E., Rayer, P., Rundle, D., Brunel, P., Vidot, J., Roquet, P., Matricardi, M., Geer, A., Bormann, N., and Lupu, C., 2018: An update on the RTTOV fast radiative transfer model (currently at version 12), Geosci. Model Dev., 11, 2717-2737, https://doi.org/10.5194/gmd-11-2717-2018.

## Installation of RTTOV
RTTOV is a free-to-use software, but the license requires each user to register to access the source code. Users should follow official instructions to access and download RTTOV v13.2 here: https://nwp-saf.eumetsat.int/site/software/rttov/download/#Software

> Once you have downloaded RTTOV, follow RTTOV's instructions to compile the libraries. Building RTTOV can take ~30 minutes, so a pre-existing build is linked to CESM for each case using COSP-RTTOV. The file "compile_RTTOV.sh" can be used as a template to compile RTTOV and download example radiative transfer coefficient files. You may need to change your compile and update paths to existing NETCDF and HDF5 libraries. NETCDF and HDF5 are used to efficiently read RTTOV coefficient files. If these libraries are not available, RTTOV can also read coefficients from .txt and binary formats, but these files are large and may take up large amounts of memeory. See RTTOV v13.2 documentation for details.

Reviewer: Typo here:

```suggestion
Once you have downloaded RTTOV, follow RTTOV's instructions to compile the libraries. Building RTTOV can take ~30 minutes, so a pre-existing build is linked to CESM for each case using COSP-RTTOV. The file "compile_RTTOV.sh" can be used as a template to compile RTTOV and download example radiative transfer coefficient files. You may need to change your compile and update paths to existing NETCDF and HDF5 libraries. NETCDF and HDF5 are used to efficiently read RTTOV coefficient files. If these libraries are not available, RTTOV can also read coefficients from .txt and binary formats, but these files are large and may take up large amounts of memory. See RTTOV v13.2 documentation for details.
```

Also is there a URL or anything to the RTTOV documentation that we could point to here?

Author: Done!

## Linking RTTOV into COSP and CAM
Once RTTOV has been built, update Makefile.rttov to point to your installation and include appropriate flags. Specifically:

> 1. Set to $RTTOVDIR to point to your installation.

Reviewer: I might be a little more specific here:

```suggestion
1. Set $RTTOVDIR to point to the directory where RTTOV is installed.
```

Author: Done!

2. If needed, update FFLAGS and LDFLAGS_ARCH to match settings used to build RTTOV.
3. If needed, update HDF5_PREFIX, FFLAGS_HDF5, LDFLAGS_HDF5, NETCDF_PREFIX, FFLAGS_NETCDF, and LDFLAGS_NETCDF to match the settings used to build RTTOV.

When building CAM/CESM, make the following changes to link RTTOV.

> 1. Set RTTOV as an environment variable (e.g. export RTTOV=TRUE, setenv RTTOV=TRUE)

Reviewer 2: @jshaw35 - Do we still need this step?

Author: I think the changes you made should remove the need for this step. I assume your test no longer explicitly sets RTTOV as an env variable? I am unable to checkout the most recent hash using gitfleximod right now to confirm from my side.

Reviewer: I also think this should no longer be a needed step (the python code in `buildlib` should do this automatically if `-rttov` is included in `CAM_CONFIG_OPTS`).

Reviewer 2: Yes, I no longer am setting an RTTOV external.  Once you are able to confirm from your side, let me know and I'll delete this line (or you can do it)

2. Add COSP and RTTOV to CAM_CONFIG_OPTS before running ./case.setup by running './xmlchange --append CAM_CONFIG_OPTS="-cosp -rttov"'
3. Run './xmlchange FORCE_BUILD_SMP=TRUE' before running ./case.setup.

## Producing satellite-like outputs

### CAM namelist settings
COSP-RTTOV allows you to specify an arbitrary number of satellite instruments to simulate. Trigger COSP-RTTOV by editing the cam namelist file (user_nl_cam) within your case. For example:

cosp_lrttov_sim = .true.,
cosp_rttov_Ninstruments = 1,

> cosp_rttov_instrument_namelists = '/glade/u/home/jonahshaw/Scripts/git_repos/COSPv2.0/driver/run/instrument_nls/cosp2_rttov_inst_CESM.txt',

Reviewer: I might update this to point to the COSP-RTTOV_examples file:
  
    
      
        Suggested change
      
    
    
      
          
            
            cosp_rttov_instrument_namelists = '/glade/u/home/jonahshaw/Scripts/git_repos/COSPv2.0/driver/run/instrument_nls/cosp2_rttov_inst_CESM.txt',
          
          
            
            cosp_rttov_instrument_namelists = '$SRCROOT/src/physics/cosp2/COSP-RTTOV_examples/cosp2_rttov_inst_CESM.txt'

Author: Done!

### Instrument coefficients.
RTTOV performs fast radiative transfer calculations using instrument-specific coefficients derived from line-by-line calculations. To simulate an instrument, coefficients must be available and you must download them to be read by COSP-RTTOV. Clear-sky calculations use Optical Depth coefficients. All-sky calculations use cloud coefficients. COSP does not handle the radiative effects of aerosols. Coefficients are available at https://nwp-saf.eumetsat.int/site/software/rttov/download/coefficients/coefficient-download/ and example code to download them is at the end of "compile_RTTOV.sh".

### Instrument namelists.
Each instrument requires a namelist (example "cosp2_rttov_inst_CESM.txt") to specify:
1. The instrument itself (e.g. AIRS, IASI) and which trace gases are used in radiative transfer calculations.
Set "rttov_srcDir" to point to your RTTOV build or the parent directory of where your RTTOV coefficients are stored.

> Set "rttov_coefDir" as the directory RTTOV coefficients are stored. "rttov_srcDir" and "rttov_coefDir" are used together when load coefficients.

Reviewer: slight grammar fix:

```suggestion
Set "rttov_coefDir" as the directory RTTOV coefficients are stored. "rttov_srcDir" and "rttov_coefDir" are used together when loading coefficients.
```

Author: Done!
